src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1504,24 +1519,133 @@ JvmtiModuleClosure::get_all_modules(JvmtiEnv* env, jint* module_count_ptr, jobje
 }
 
 void
-UpdateForPopTopFrameClosure::do_thread(Thread *target) {
+UpdateForPopTopFrameClosure::doit(Thread *target, bool self) {
+  Thread* current_thread  = Thread::current();
+  HandleMark hm(current_thread);
+  JavaThread* java_thread = target->as_Java_thread();
```

Comment:
Contrast with the same three lines at L1390 - we should use the same boilerplate in each `doit`. And ideally refactor into some shared code somewhere (future RFE).

Reply:
Yes, that would be good.

Reply:
Why don't you just do:
JavaThread* java_thread = JavaThread::current();
HandleMark hm(java_thread);

JavaThread::current is the same thing as what you have.

Reply:
Hi Coleen, we are in a handshake.
Target java_thread must be treat as different from current thread. 
It can be same, but also another thread java thread or VM thread.
So we only know the the thread executing is a Thread, but our target is a JavaThread.

Here we want to create a HandleMark in current thread.
Then we want a JavaThread pointer to the target for later use.

Does that explain it?

Reply:
I see it now, different threads.